### DISCUSSION TITLE
Store content in sessionStorage

### DISCUSSION THREAD

**Discussion by miquik** [pmndrs/jotai#2265] [Q&A]
Hi all,
I am using several atoms in my app. I'm thinking that if the user inadvertently presses the reload button, everything would be reset.
I know it is possible to use atomWithStorage, but what if I have many atoms? Do I have to create them all this way?
Couldn't you do something like save the store in the session?

Thanks

**Comment by dai-shi** [comment]
Jotai store is a WeakMap. There's no way to know "all atoms" in the store. So, it's impossible by design.

Technically, you could create a "catch-all" atom and store it to the storage.

**Comment by himself65** [comment]
I encountered the same issue before when my app was running on different platforms, so there are different kinds of storage.

So my solution is to add hooks on the atom, and I publish a package called `jotai-inject`

Here is the source code you can copy/paste

```ts
export function inject<
  Value
> (
  injectAtom: PrimitiveAtom<Value> & { init: Value | Promise<Value> },
  accessor: () => (Value | Promise<Value>),
  updateValue?: (apply: Value) => void
) {
  const originalWrite = injectAtom.write.bind(injectAtom)
  injectAtom.init = accessor()
  injectAtom.write = function (get, set, apply) {
    originalWrite(get, set, apply)
    updateValue?.(get(injectAtom))
    return
  }
}
```

The usage:

```ts
inject(userInfoAtom, () => {
  localStorage.get(''xxx')
}, (value) => {
  localStorage.set('xxx', JSON.stringify(value))
})
```

Please call it at the top of the app.
